File: node/src/tile.ts
import {
  nextTileMap,
  Suit,
  suitCodeMap,
  UnicodeTileLookup,
  valueCodeMap,
} from './constants/tiles';

import { type TileValue, type TileString } from './types/tile';
import { type ValueOf } from './types/util';

export class Tile {
  suit: Suit;
  value: TileValue;
  name: TileString;
  unicode: ValueOf<typeof UnicodeTileLookup>;

  public isAkadora: boolean = false;
  public isDora: boolean = false;
  public isWild: boolean = false;

  constructor(suit: keyof typeof Suit, value: TileValue) {
    this.suit = suit;
    this.value = value;

    this.name = `${value}_${suit}` as TileString;
    this.unicode = UnicodeTileLookup[this.name];
  }

  public get isNumber(): boolean {
    return typeof this.value === 'number';
  }
  public get isTerminal(): boolean {
    return (
      typeof this.value === 'number' && (this.value === 1 || this.value === 9)
    );
  }
  public get isSimple(): boolean {
    return typeof this.value === 'number' && (this.value > 1 || this.value < 9);
  }
  public get isBonus(): boolean {
    return this.suit === Suit.SEASON || this.suit === Suit.FLOWER;
  }
  public get isHonor(): boolean {
    return this.suit === Suit.WIND || this.suit === Suit.DRAGON;
  }
  public get isGreen(): boolean {
    return (
      this.value === 'GREEN' ||
      (this.suit === Suit.BAMBOO &&
        (this.value === 2 ||
          this.value === 3 ||
          this.value === 4 ||
          this.value === 6 ||
          this.value === 8))
    );
  }

  public to_code(): string {
    if (this.isWild) {
      return '0j';
    }

    const suitCode = suitCodeMap[this.suit];
    if (this.isNumber) {
      const value = this.isAkadora ? 0 : this.value;
      return `${value}${suitCode}`;
    }

    if (this.isHonor || this.isBonus) {
      const value = valueCodeMap[this.name as keyof typeof valueCodeMap];
      return `${value}${suitCode}`;
    }

    return '';
  }

  public static from_string(name: string): Tile {
    const [value, suit, ..._] = name.split('_');
    return new Tile(suit as Suit, value as TileValue);
  }

  public static next(tile: Tile, step: number = 1): Tile {
    const nextTileName = nextTileMap[tile.name];

    return Tile.from_string(nextTileName);
  }
}

File: rust/src/structs/tile.rs
use std::{
    cmp::Ordering,
    fmt::{self, Display},
    hash::Hash,
    str::FromStr,
};

use derivative::Derivative;

use crate::{
    consts::TileData,
    maps::NEXT_TILE_MAP,
    notation::{tile_parse_error::TileParseError, trait_tile_code::TILE_CODE_MAP},
};
use crate::{
    consts::{tile::TileNumber, Dragon, TileName},
    notation::ToTileCode,
};

#[derive(Derivative, Debug)]
#[derivative(Default)]
pub struct Tile {
    pub tile_data: TileData,
    pub name: TileName,

    #[derivative(Default(value = "false"))]
    pub is_wild: bool,
    #[derivative(Default(value = "false"))]
    pub is_dora: bool,
    #[derivative(Default(value = "false"))]
    pub is_akadora: bool,
}

impl Tile {
    pub fn new(tile_data: TileData) -> Tile {
        let name = TileName::from(&tile_data);

        Tile {
            tile_data,
            name,
            ..Default::default()
        }
    }

    pub fn is_number(&self) -> bool {
        matches!(
            self.tile_data,
            TileData::BAMBOO(_) | TileData::MAN(_) | TileData::TONG(_)
        )
    }

    pub fn is_terminal(&self) -> bool {
        self.is_number()
            && match &self.tile_data {
                TileData::BAMBOO(value) | TileData::MAN(value) | TileData::TONG(value) => {
                    matches!(value, TileNumber::ONE | TileNumber::NINE)
                }
                _ => false,
            }
    }

    pub fn is_simple(&self) -> bool {
        self.is_number() && !self.is_terminal()
    }

    pub fn is_honor(&self) -> bool {
        matches!(self.tile_data, TileData::WIND(_) | TileData::DRAGON(_))
    }

    pub fn is_bonus(&self) -> bool {
        matches!(self.tile_data, TileData::SEASON(_) | TileData::FLOWER(_))
    }

    pub fn is_green(&self) -> bool {
        matches!(
            self.tile_data,
            TileData::DRAGON(Dragon::GREEN)
                | TileData::BAMBOO(
                    TileNumber::TWO
                        | TileNumber::THREE
                        | TileNumber::FOUR
                        | TileNumber::SIX
                        | TileNumber::EIGHT
                )
        )
    }
}

impl ToTileCode for Tile {
    fn to_tile_code(&self) -> String {
        if self.is_wild {
            return String::from("0j");
        }

        let tile_code_str = TILE_CODE_MAP[&self.name];

        if self.is_number() && self.is_akadora {
            return String::from("0") + &tile_code_str[1..];
        }

        String::from(tile_code_str)
    }
}

impl Display for Tile {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.name)
    }
}

impl FromStr for Tile {
    type Err = TileParseError;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let tile_data = TileData::try_from(s.to_owned());
        if tile_data.is_err() {
            return Err(tile_data.err().unwrap());
        }

        Ok(Tile::new(tile_data.unwrap()))
    }
}

impl PartialEq for Tile {
    fn eq(&self, other: &Self) -> bool {
        self.name == other.name
    }
}
impl Eq for Tile {}

impl Hash for Tile {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.name.hash(state);
    }
}

impl PartialOrd for Tile {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        let left_index = NEXT_TILE_MAP.get_index_of(&self.name);
        let right_index = NEXT_TILE_MAP.get_index_of(&other.name);

        Some(if left_index > right_index {
            Ordering::Greater
        } else if right_index < left_index {
            Ordering::Less
        } else {
            Ordering::Equal
        })
    }
}
